postgis/lwgeom_remove_irrelevant_points_for_view.c

```diff
@@ -128,8 +128,20 @@ static int encodeToBitsStraight(double xa, double ya, double xb, double yb, doub
 // outside points would be optimal to keep. Since this would introduce a lot
 // more code complexity and a backing array and would likely have
 // no real practical impact this step is skipped.
+//
+// Note on cartesian_hint:
+// - if false, the algorithm removes one or a sequence of points
+//   lying on "the same side" (either top, bottom, left or right) of the
+//   given bounds except the first and last point of that sequence.
+// - if true, the algorithm assumes that the coordinates are rendered in
+//   a cartesian coordinate system and tries to remove further points
+//   if the resulting connection lines do not cross the borders of
+//   the rectangular view given by the bounds.
+//   Please note that this option might produce rendering artifacts
+//   if the coordinates are used for rendering in a non-cartesian
+//   coordinate system.
 // ===============================================================================
-static void removePoints(POINTARRAY *points, GBOX *bounds, bool closed) {
+static void removePoints(POINTARRAY *points, GBOX *bounds, bool closed, bool cartesian_hint) {
```

Comment:
Any reason why this is not in liblwgeom  ? Moving it there would allow adding unit tests for proper memory management debugging

Reply:
No, actually no special reason. What is the difference between _postgis_ and _liblwgeom_, and are there other files to be adapted when moving _lwgeom_remove_irrelevant_points_for_view.c_ from _postgis_ to _liblwgeom_ folder? (Same question for _lwgeom_st_remove_small_parts.c_?)

Reply:
@strk When I move the .c file to _liblwgeom_, adapting _postgis/Makefile.in_ and _liblwgeom/Makefile.in_ I get a lot of compiler errors because of missing import statements. I'm not sure how I could resolve them all. Could you please assist me, or would it be an option to leave the file at its current location?

Reply:
show the errors ? You mean include statement, I guess. The usefulness of moving the code under liblwgeom is for being able to write unit tests which usually catch more errors, especially memory management related.

Reply:
@strk You are right, I meant _include_, not _import_. The errors I get when compiling in _/liblwgeom_ are:
```
fatal error: postgres.h: No such file or directory
fatal error: funcapi.h: No such file or directory
fatal error: utils/array.h: No such file or directory
fatal error: utils/builtins.h: No such file or directory
fatal error: utils/lsyscache.h: No such file or directory
fatal error: utils/numeric.h: No such file or directory
fatal error: access/htup_details.h: No such file or directory
```
If I knew how to get rid of these errors I'd try adding cunit tests.

Since IMO there's nothing crucial related to memory management in my code - could we open a new ticket for this refactoring step? I'd like to do the same for _ST_RemoveSmallParts_. What do you think?

Reply:
The functionality needs be split in two.
One part will only touch LWGEOM object and won't need any of those headers.
The other part will be just a thin wrapper for that liblwgeom function, to expose it to SQL, and only that part will need some of those headers (not all, most likely)

Reply:
> The functionality needs be split in two.

@strk: I splitted the functionality into two parts (.c and .h in _/liblwgeom_ without the above includes, and the shortened .c in _/postgis_). Could you please have a look at it?

What else is required? I could add _CUnit_ tests, but when I run _make_ in _liblwgeom/cunit_ I get "configure was unable to find CUnit which is required for unit testing." although _CUnit_ should be installed. What could be the mistake here?

Reply:
@strk: It would be great if this PR could be merged before the upcoming 3.5.0 release since using the function on non-cartesian coordinates can lead to wrong results in some cases as explained in the trac ticket. By introducing the _cartesian_hint_ flag this possible source of error is clearly removed. I'm not sure if it would be clearer to tag this PR / trac ticket as "defect" instead of "enhancement".

Reply:
Could someone (@strk?) give a short update on this please? I look forward that this fix will get it into 3.5.0 along the introduction of ST_RemoveIrrelevantPointsForView(). Thank you.

Reply:
I would take one step further and expose an lwgeom_remove_irrelevant_points_for_view() function, with its unit test.

Reply:
@strk: With the latest commit I added lwgom_remove_irrelevant_points_for_view(). Is this ok?
A unit test will follow, I first have to get CUnit working...

Reply:
Ok, got CUnit working now and added a working unit test for lwgeom_remove_irrelevant_points_for_view().

Reply:
@strk: Thank you for another review :) It would be great to get this fix still 3.5.0 (fixed for the 1st of September - https://trac.osgeo.org/postgis/roadmap)?